src/scanner.c

```diff
@@ -0,0 +1,205 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
+
+static String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
+}
+
+static void string_resize(String *vec, uint32_t cap) {
+    void *block = realloc(vec->data, (cap + 1) * sizeof(vec->data[0]));
+    assert(block != NULL);
```

Comment:
`assert` is a debug macro, can be removed by compiler in optimized release code,
I think explicitly `exit(EXIT_FAILURE);` is enough.